slo-monitor/src/monitors/pod_monitor.go

```diff
@@ -0,0 +1,356 @@
+/*
+Copyright 2017 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package monitors
+
+import (
+	"strings"
+	"sync"
+	"time"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/wait"
+	"k8s.io/apimachinery/pkg/watch"
+
+	"k8s.io/apiserver/pkg/storage"
+
+	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/api/v1"
+	"k8s.io/client-go/tools/cache"
+
+	kubeletevents "k8s.io/kubernetes/pkg/kubelet/events"
+
+	"github.com/golang/glog"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const (
+	// PodE2EStartupLatencyKey is a key for pod startup latency monitoring metric.
+	PodE2EStartupLatencyKey = "slomonitor_pod_e2e_startup_latency_seconds"
+	// Time after which deleted entries are purged
+	purgeAfter = 2 * time.Minute
+)
+
+var (
+	// PodE2EStartupLatency is a prometheus metric for monitoring pod startup latency.
+	PodE2EStartupLatency = prometheus.NewHistogram(
+		prometheus.HistogramOpts{
+			Name:    PodE2EStartupLatencyKey,
+			Help:    "Pod e2e startup latencies in milliseconds, without image pull times",
+			Buckets: prometheus.LinearBuckets(0.5, 0.25, 50),
+		},
+	)
+)
+
+var registerMetrics sync.Once
+
+// Register registers all prometheus metrics.
+func Register() {
+	registerMetrics.Do(func() {
+		prometheus.MustRegister(PodE2EStartupLatency)
+	})
+}
+
+// PodStartupMilestones keeps all milestone timestamps from Pod creation.
+type PodStartupMilestones struct {
+	created          time.Time
+	startedPulling   time.Time
+	finishedPulling  time.Time
+	observedRunning  time.Time
+	deletedTimestamp time.Time
+}
+
+// PodStartupLatencyDataMonitor monitors pod startup latency and exposes prometheus metric.
+type PodStartupLatencyDataMonitor struct {
+	sync.Mutex
+	kubeClient     clientset.Interface
+	PodStartupData map[string]PodStartupMilestones
+	toDelete       []string
+	startPodRV     uint64
+	startEventRV   uint64
```

Comment:
I'm using those to double check if we're not getting something too old in 'handleX' functions. I can remove them if you think that's unnecessary.